includes/class-amp-story-post-type.php

```diff
@@ -646,4 +721,188 @@ public static function the_single_story_card( $post ) {
 		</a>
 		<?php
 	}
+
+	/**
+	 * Enqueues this post type's stylesheet for the embed endpoint and Latest Stories block.
+	 */
+	public static function enqueue_embed_styling() {
+		if ( is_embed() && is_singular( self::POST_TYPE_SLUG ) ) {
+			wp_enqueue_style( self::STORY_CARD_CSS_SLUG );
+		}
+	}
+
+	/**
+	 * Overrides the render_callback of an AMP story post embed, when using the WordPress (embed) block.
+	 *
+	 * WordPress post embeds are usually wrapped in an <iframe>,
+	 * which can cause validation and display issues in AMP.
+	 * This overrides the embed callback in that case, replacing the <iframe> with the simple AMP story card.
+	 *
+	 * @param string $pre_render The pre-rendered markup, default null.
+	 * @param array  $block The block to render.
+	 * @return string|null $rendered_markup The rendered markup, or null to not override the existing render_callback.
+	 */
+	public static function override_story_embed_callback( $pre_render, $block ) {
+		if ( ! isset( $block['attrs']['url'], $block['blockName'] ) || ! in_array( $block['blockName'], array( 'core-embed/wordpress', 'core/embed' ), true ) ) {
+			return $pre_render;
+		}
+
+		// Taken from url_to_postid(), ensures that the URL is from this site.
+		$url           = $block['attrs']['url'];
+		$url_host      = wp_parse_url( $url, PHP_URL_HOST );
+		$home_url_host = wp_parse_url( home_url(), PHP_URL_HOST );
+
+		// Exit if the URL isn't from this site.
+		if ( $url_host !== $home_url_host ) {
+			return $pre_render;
+		}
+
+		$path = str_replace( home_url( self::REWRITE_SLUG . '/' ), '', $url );
```

Comment:
This still doesn't seem very robust. What if the `$url` is an HTTP one, but the site is in HTTPS? Since you've already verified above that the host is the same, it seems this should rather do something like:

```suggestion
		$embed_url_path = wp_parse_url( $url, PHP_URL_PATH );
		$base_url_path  = wp_parse_url( trailingslashit( home_url( self::REWRITE_SLUG ) ), PHP_URL_PATH );
		if ( 0 !== strpos( $embed_url_path, $base_url_path ) ) {
			return $pre_render;
		}
		$path = substr( $embed_url_path, strlen( $base_url_path ) );
```

Reply:
Thanks, that's a good point. I [committed](https://github.com/ampproject/amp-wp/pull/1944/commits/80b3a7ea6558ec9b4c8a7bf71267846110a0d593) your code verbatim.